### ISSUE TITLE
Create an AMI by importing virtual machine images from any virtualization environment

### ISSUE THREAD

**Issue by mchouque** [hashicorp/terraform-provider-aws#5998]
Hello,

With EC2 on AWS, you can use AMIs provided by Amazon or you can "bring your own OS".

Namely you can (it's documented here [Importing a VM as an Image Using VM Import/Export](https://docs.aws.amazon.com/vm-import/latest/userguide/vmimport-image-import.html):)
- upload a copy of an existing OS disk image you have to a S3 bucket with "_aws s3 cp myimage_" (can be a vhd, vmdk or ova file)
- trigger a conversion to an AMI using "_aws ec2 import-image_"

When it's done, you have an AMI that was created with the disk image you provided: your own OS tailored to your needs.

In a nutshell, with google, the code to do that is as follow (don't mind the crappy variable names and code, I'm just starting using terraform):
```hcl
resource "google_storage_bucket_object" "linux-image" {
  name   = "${var.file_name}"
  source = "${var.image_path}/${var.file_name}"
  bucket = "${google_storage_bucket.linux-bucket.name}"
}

resource "google_compute_image" "linux-image" {
  name        = "${var.image_name}"
  family      = "${var.image_family}"
  description = "${var.image_description}"

  raw_disk {
    source = "https://storage.googleapis.com/${var.bucket_name}/${google_storage_bucket_object.linux-image.name}"
  }
}
```
Namely two resources: one to upload a file to a bucket, the other to "transform" it.

With Azure, it's very similar: 2 resources. One to upload the image, and one to convert it.
```hcl
resource "azurerm_storage_blob" "linux-image" {
  name                   = "${var.image_name}"
  resource_group_name    = "${var.resource_group}"
  storage_account_name   = "${azurerm_storage_account.linux-bucket.name}"
  storage_container_name = "${azurerm_storage_container.linux-container.name}"
  type                   = "page"
  source                 = "${var.image_path}/${var.image_name}"
}

resource "azurerm_image" "linux-image" {
  name                = "${var.image_name_short}"
  resource_group_name = "${var.resource_group}"
  location            = "${var.bucket_location}"

  os_disk {
    os_type  = "Linux"
    blob_uri = "${azurerm_storage_blob.linux-image.url}"
  }
}
```

With Amazon, we have the first resource, it's "_aws_s3_bucket_object_".

The only thing missing in terraform for AWS is the equivalent of the second resource, azurerm_image for Azure or google_compute_image for Google.

On EC2, the conversion is triggered by "_aws ec2 import-image_" and I believe the API call for that is [ImportImage](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ImportImage.html).

Let me know if you have any questions.

Regards,
Mathieu

**Comment by nealej81** [comment]
Is there any way of doing this yet with Terraform in AWS?

**Comment by zzzuzik** [comment]
same question here, is it possible now?

**Comment by penguin-27** [comment]
I would love to see a Terraform resource for this as well!

**Comment by artificial-aidan** [comment]
> This is possible now through a combination of `aws_ebs_snapshot_import` and `aws_ami`.

This is not true for all cases. You need image import, not snapshot import for a number of use cases.

I started https://github.com/hashicorp/terraform-provider-aws/pull/25621 to try and support this.

**Comment by alexeiser** [comment]
> > This is possible now through a combination of `aws_ebs_snapshot_import` and `aws_ami`.
> 
> This is not true for all cases. You need image import, not snapshot import for a number of use cases.
> 
> I started #25621 to try and support this.

Hopefully they will accept your MR - out of curiosity what use case does the import handle that the snapshot import and custom ami creation step not handle?

**Comment by artificial-aidan** [comment]
When I imported a Windows VM disk, then created an AMI, a Linux AMI was created. `aws_ami` doesn't allow platform specification.

**Comment by trevor-viljoen** [comment]
Is this enhancement still actively being considered? Is there an ETA on when it could be implemented if so? I need to create an EC2 from an OVA appliance that was sent to us by the vendor and the preferred way of doing that would be with Terraform.
